src/os/bluestore/BlueStore.h

```diff
@@ -1547,15 +1547,27 @@ class BlueStore : public ObjectStore,
 
     std::atomic_int kv_submitted_waiters = {0};
 
+    std::mutex register_lock;
+    bool registered = true;
+
     OpSequencer(CephContext* cct, BlueStore *store)
 	//set the qlock to PTHREAD_MUTEX_RECURSIVE mode
       : Sequencer_impl(cct),
 	parent(NULL), store(store) {
+      std::lock_guard<std::mutex> l(register_lock);
       store->register_osr(this);
     }
     ~OpSequencer() {
       assert(q.empty());
-      store->unregister_osr(this);
+      discard();
+    }
+
+    void discard() override {
+      std::lock_guard<std::mutex> l(register_lock);
+      if (registered) {
```

Comment:
Still disagree.. There is a lock within (un)register_osr and hence osr_set is protected by itself. OpSequencer ctor doesn't need mutex and can simply set registered to true and call protected register_osr. The only questionable location is discard. If it's called from different threads simultaneously the luckiest one atomically clears 'registered' var via atomic::compare_and_exchange and proceeds to unregister call. The loser just compares and do nothing.  There is no need for additional mutex here if registered is std::atomic<bool> instance.